src/interface/MediaSource.js

```diff
@@ -0,0 +1,140 @@
+let audioElement = null;
+
+// A reference to an HTML5 `MediaSource` instance.
+let source = null;
+
+// A reference to the internal MPEG buffer for the `source`. Appending MPEG data
+// to this buffer results in it being played.
+let sourceBuffer = null;
+
+// While playing, the system uses a recursive `setTimeout` loop to continuously
+// append audio data to the `sourceBuffer`. This is a reference to that timeout
+// and only exists while audio is being played.
+let sourceTimeout = null;
+
+function initialize({ volume = 1.0 }) {
+  audioElement = document.querySelector('audio');
+  source = new MediaSource();
+  sourceBuffer = null;
+  sourceTimeout = null;
+
+  setVolume(volume);
+}
+
+function play(clip) {
+  source = new MediaSource();
+  source.addEventListener('sourceopen', function () {
+    console.log('source open');
+    sourceBuffer = this.addSourceBuffer('audio/mpeg');
+    _play(clip);
+  });
+
+  audioElement.src = URL.createObjectURL(source);
+
+  return audioElement.play().catch(suppressAbortError);
+}
+
+function _play(clip) {
+  if (clip._clipState.chunksBufferingFinished) {
+    debug('MediaSourceEngine endOfStream()');
+    source.endOfStream();
+    return;
+  }
+
+  const isChunkReady = clip._clipState.isChunkReady(clip._clipState.chunkIndex);
+
+  const useSilence =
+    !isChunkReady &&
+    clip._clipState.chunkIndex === clip._initialChunk &&
+    !clip._wasPlayingSilence &&
+    (clip.browserName === 'safari' || clip.osName === 'ios');
+
+  const chunk = useSilence
+    ? clip._silenceChunks[0]
+    : isChunkReady && clip._clipState.chunks[clip._clipState.chunkIndex];
+
+  if (chunk) {
+    try {
+      sourceBuffer.appendBuffer(chunk.raw);
+      if (isChunkReady) {
+        clip._clipState.chunkIndex += 1;
+        clip._wasPlayingSilence = false;
+      } else if (useSilence) {
+        clip._wasPlayingSilence = true;
+      }
+    } catch (e) {
+      // SourceBuffer might be full, remove segments that have already been played.
+      debug('Exception when running SourceBuffer#appendBuffer', e);
+      try {
+        sourceBuffer.remove(0, audioElement.currentTime);
+      } catch (e) {
+        debug('Exception when running SourceBuffer#remove', e);
+      }
+    }
+  }
+
+  const timeout = isChunkReady ? Math.min(500, chunk.duration * 1000) : 100;
+  sourceTimeout = setTimeout(() => _play(clip), timeout);
+}
+
+function resume() {
+  return audioElement.play().catch(suppressAbortError);
+}
+
+function pause() {
+  audioElement.pause();
+}
+
+function stop() {
+  pause();
+
+  clearTimeout(sourceTimeout);
+  sourceTimeout = null;
+}
+
+function setVolume(volume) {
+  audioElement.volume = volume;
+}
+
+function isPlaying() {
+  return getPlaybackState() === PlaybackState.Playing;
+}
+
+function getPlaybackState() {
+  if (audioElement.paused) return PlaybackState.Paused;
+  if (sourceBuffer === null) return PlaybackState.Stopped;
+  return PlaybackState.Playing;
+}
+
+function __TEMP__currentTime(clip) {
```

Comment:
Why `__TEMP__`? Feel free to answer just one of these, if the answer is the same for all 😝

Reply:
This is just a reminder to me to remove these. I tend to put things like `__DO_NOT_USE_THIS_PLEASE__`, `__TEMP__`, etc, directly in the function name of things that I plan on removing. Putting this info in comments never actually works.

`currentTime` will be removed in the future because the `Clip` construct will just be a container for track information and won't have any concept of the player's state.